apis/fluentbit/v1alpha2/clustermultilineparser_types.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/fluent/fluent-operator/v2/apis/fluentbit/v1alpha2/plugins"
+	"github.com/fluent/fluent-operator/v2/apis/fluentbit/v1alpha2/plugins/multilineparser"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"reflect"
+	"sort"
+)
+
+// +kubebuilder:object:generate:=true
+
+type MultilineParserSpec struct {
+	MultilineParser *multilineparser.MultilineParser `json:",inline"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:shortName=cfbmp,scope=Cluster
+// +genclient
+// +genclient:nonNamespaced
+
+// ClusterMultilineParser is the Schema for the cluster-level multiline parser API
+type ClusterMultilineParser struct {
```

Comment:
Why need to create a new CRD? I think multiline Parser is also a kind of `Parser`, it can use Parser CRD.

Reply:
1. the section for multiline parser is `[MULTILINE_PARSER]` instead `[PARSER]` and they're treated as the same top level
2. multiline parser is not located under Parsers in fluentbit doc

i agree it's better to treat multiline parser as one kind of parser as well.

but it seems like currently it's not implemented like this from fluentbit's perspective. it's also reasonable to follow the convention of fluentbit so that customers won't get confused.

Reply:
Got it.

Does it need to add the `parsers_multiline.conf` to the `SERVER` section?

<img width="522" alt="1690248036164" src="https://github.com/fluent/fluent-operator/assets/53003665/63526405-a230-40bb-9df5-3ece25cae261">

Reply:
yes. so i propose to add one new field called `parsersFiles` to support multiple `parsers_file` in `SERVICE` section, and keep `parsersFile` for backward compatibility: https://github.com/fluent/fluent-operator/pull/842/files#diff-7aad9bf9eeeaa93f3a61c5ea93ea5a30d57bcfa6307ec4ea188222fdf69f438cR180 in case customers want to add parsers.conf and parsers_multiline.conf at the same time.

Reply:
@wanjunlei would you review this again?

Reply:
> backward compatibility

Does this mean the user needs to manually set `parsersFiles` after defining a `MultilineParser`?

Reply:
actually no. this field is used when customers want to specify multiple parser files, such as one with `[PARSER]` section and one with `[MULTILINE_PARSER]` section.

Reply:
i'll test this tomorrow.

Reply:
1. [x] fix rbac 
2. [ ] inline for `MultilineParser`

Reply:
> how to set the ClusterFluentBitConfig when I define a MultilineParser and want to use it

```yaml
apiVersion: fluentbit.fluent.io/v1alpha2
kind: ClusterFluentBitConfig
metadata:
  labels:
    app.kubernetes.io/name: fluent-bit
  name: fluent-bit-config
spec:
  filterSelector:
    matchLabels:
      fluentbit.fluent.io/enabled: "true"
  inputSelector:
    matchLabels:
      fluentbit.fluent.io/enabled: "true"
  multilineparserSelector:
    matchLabels:
      fluentbit.fluent.io/enabled: "true"
  outputSelector:
    matchLabels:
      fluentbit.fluent.io/enabled: "true"
  parserSelector:
    matchLabels:
      fluentbit.fluent.io/enabled: "true"
  service:
    logLevel: info
    parsersFile: parsers_multiline.conf
```

```yaml
apiVersion: fluentbit.fluent.io/v1alpha2
kind: ClusterMultilineParser
metadata:
  labels:
    fluentbit.fluent.io/component: logging
    fluentbit.fluent.io/enabled: "true"
  name: stacktrace
spec:
  flushTimeout: 1000
  rules:
  - next: cont
    regex: /([a-zA-Z]+ \d+ \d+\:\d+\:\d+)(.*)/
    start: start_state
  - next: cont
    regex: /^\s+at.*/
    start: cont
  type: regex
```

Reply:
By default, the `parsersFile` in `ClusterFluentBitConfig` is `parsers.conf`, And it points to [/fluent-bit/etc/parsers.conf](https://github.com/fluent/fluent-operator/blob/master/conf/parsers.conf) which include the common parsers such as `docker`, `cri`. If it was set to `parsers_multiline.conf`, the fluentbit cannot find these parsers.

BTW, if the `parsersFile` in `ClusterFluentBitConfig` is a relative path, the fluentbit will find it in the directory where the fluentbit configuration file is located, which is `/fluent-bit/etc/`, but the `parsers_multiline.conf` is in `/fluent-bit/config/`.

Maybe you should handle the `parsers_multiline.conf` like `parsers.conf`

<img width="924" alt="企业微信截图_16913732558404" src="https://github.com/fluent/fluent-operator/assets/53003665/8b98ec68-d18e-498b-a3d9-f482df31c57f">